samples/subsys/shell/shell_module/src/uart_reinit.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <shell/shell.h>
+#include <shell/shell_uart.h>
+#include <drivers/uart.h>
+#include <device.h>
+
+void shell_init_from_work(struct k_work *work)
+{
+	const struct device *dev =
+			device_get_binding(CONFIG_UART_SHELL_ON_DEV_NAME);
+	bool log_backend = CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL > 0;
+	uint32_t level =
+		(CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL > LOG_LEVEL_DBG) ?
+		CONFIG_LOG_MAX_LEVEL : CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL;
+
+	shell_init(shell_backend_uart_get_ptr(), dev, true, log_backend, level);
+}
+
+static void shell_reinit_trigger(void)
+{
+	static struct k_work shell_init_work;
+
+	k_work_init(&shell_init_work, shell_init_from_work);
+	int err = k_work_submit(&shell_init_work);
+
+	(void)err;
+	__ASSERT_NO_MSG(err >= 0);
+}
+
+static void direct_uart_callback(const struct device *dev, void *user_data)
```

Comment:
What if `CONFIG_UART_INTERRUPT_DRIVEN` isn't defined, this function seems to heavily rely on interrupts. Could we get away with calling poll_in and poll_out instead?

Reply:
I've updated sample to support case when shell uart is not using interrupts.